### PR TITLE
Xin add l0

### PR DESCRIPTION
**Pull request by XinWangIntel**
Add L0 draft

### CODE REVIEW

**Comment by XinWangIntel** [comment]
Can pass resnet50 with graphic card(630) and compute-runtime(20.40.18075), but failed with graphic card(650) with compute-runtime( 20.37.0), driver version matters.

File: plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+CppEdsl.HigherPrecisionConstants
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:1]: Are all these tests still not pass?

File: plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+CppEdsl.SinH
+CppEdsl.Tan
+CppEdsl.ConvI8
+CppEdsl.InterpolatedGatherNearest
+CppEdsl.InterpolatedGatherLinear
+CppEdsl.InterpolatedGatherCubic
+CppEdsl.InterpolatedGatherCubicDouble
+CppEdsl.InterpolatedGatherCubicHalf
+CppEdsl.InterpolatedGatherMultiDIndices
+CppEdsl.BitNot
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt]: This one is redundant.

File: pmlc/BUILD [hunk 3]
```diff
@@ -25,9 +25,11 @@ filegroup(
         "//pmlc/dialect/tile/tests:all",
         "//pmlc/dialect/xsmm/tests:all",
         "//pmlc/rt/opencl/tests:all",
+        "//pmlc/rt/levle_zero/tests:all",
```
Comment (YangleiZouIntel) [inline pmlc/BUILD]: Typo: level_zero

File: pmlc/rt/level_zero/.gdb_history [hunk 4]
```diff
@@ -0,0 +1,8 @@
+where
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/.gdb_history]: This file is accidently added.

File: pmlc/conversion/comp_to_llvm/passes.td [hunk 5]
```diff
@@ -17,4 +17,13 @@ def ConvertCompToLLVM : Pass<"pmlc-convert-comp-to-llvm", "mlir::ModuleOp"> {
   ];
 }
 
+def ConvertCompToLevelZero : Pass<"pmlc-convert-comp-to-level-zero", "mlir::ModuleOp"> {
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/passes.td]: This should be deleted.

File: pmlc/conversion/comp_to_llvm/passes.h [hunk 6]
```diff
@@ -33,6 +33,24 @@ std::unique_ptr<mlir::Pass> createConvertCompToLLVMPass();
 std::unique_ptr<mlir::Pass>
 createConvertCompToLLVMPass(const std::string &prefix);
 
+// ============================================================================
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/passes.h]: This should be deleted.

File: pmlc/rt/level_zero/level_zero_device.cc [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+#include "pmlc/rt/jit_executable.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::rt::level_zero {
+
+LevelZeroQueue::LevelZeroQueue(const ze_context_handle_t &context,
+                               const ze_device_handle_t &device,
+                               ze_command_queue_group_properties_t properties)
+    : properties(properties) {
+  queue = lzu::create_command_queue(context, device, 0,
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_device.cc]: You might want to add llvm style comments `/*description*/ 0, ` for these 0

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+  lzu::set_argument_value(kernel, idx, sizeof(buf), &buf);
+}
+
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // Need to find a better way to find right group count.
+  lzu::append_launch_function(list, kernel, &gws, resultE, dependencies.size(),
+                              dependencies.data());
+  // As changes in wrapper, do not need to divide now.
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc]: You can delete these comments now.

File: pmlc/rt/level_zero/level_zero_device.cc [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                                 ze_device_handle_t device)
+    : driver(driver), device(device) {
+  context = lzu::get_context(driver);
+  IVLOG(1, "Instantiating LevelZero device: "
+               << lzu::get_device_properties(device).name);
+  queues.clear();
+}
+
+LevelZeroDevice::~LevelZeroDevice() {
+  // Once multiple devices, shall change
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_device.cc]: This comments needs some rewording and other functions will need changes as well.  
In addition I think we will need one more abstraction level/interface for multi-devices.

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+  finish();
+
+  for (size_t i = 0; i < memories.size(); i++) {
+    delete memories[i];
+  }
+  for (std::unique_ptr<LevelZeroEvent> &event : events) {
+    eventPool.destroy_event(event->getEvent());
+  }
+  // for (size_t i = 0; i < kernels.size(); i++) {
+  //  delete kernels[i];
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc]: kernels are not used now?

File: pmlc/rt/level_zero/level_zero_invocation.h [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+  /// Blocks untill all previously enqueued operations are finished.
+  void finish();
+
+private:
+  std::shared_ptr<LevelZeroDevice> device;
+  LevelZeroQueueUser queueUser;
+  lzu::zeEventPool eventPool;
+  // clear list
+  std::vector<std::unique_ptr<LevelZeroEvent>> events;
+  std::vector<LevelZeroKernel *> kernels;
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.h]: `kernels` is not used?

File: pmlc/rt/level_zero/level_zero_invocation.h [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  std::vector<LevelZeroMemory *> memories;
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.h:145]: <del>Maybe it is better to use `std::vector<std::shared_ptr<LevelZeroMemory>>` so we do not need to manually free the memory.

Ok, this vector contains all the memories that explicitly deallocated, I think we can keep it as it is.

File: pmlc/rt/level_zero/level_zero_invocation.h [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+  /// blocked by any dependencies.
+  void flush();
+  /// Blocks untill all previously enqueued operations are finished.
+  void finish();
+
+private:
+  std::shared_ptr<LevelZeroDevice> device;
+  LevelZeroQueueUser queueUser;
+  lzu::zeEventPool eventPool;
+  // clear list
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.h]: <del>What does this comment mean?

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    zeEventHostSynchronize(e, UINT64_MAX);
+  }
+}
+ze_command_queue_group_properties_t p;
+LevelZeroInvocation::LevelZeroInvocation(LevelZeroDevice *device)
+    : device{device->shared_from_this()}, queueUser(device->getQueue(p)) {
+  // Need a way to ensure event is not used and then can release it
+  eventPool.InitEventPool(device->getLevelZeroContext(), 600);
+
+  // May create queue based on properties once multiple device used
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc]: This comment needs some rewording.

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+                                                         size_t bytes,
+                                                         const char *name) {
+  uint8_t *buf = reinterpret_cast<uint8_t *>(data);
+  ze_module_handle_t module = lzu::create_module(
+      device->getLevelZeroContext(), device->getLevelZeroDevice(), buf, bytes,
+      ZE_MODULE_FORMAT_IL_SPIRV, "", nullptr);
+
+  LevelZeroKernel *kernel = new LevelZeroKernel(module, name);
+  // kernels.push_back(kernel);
+  return kernel;
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc]: `return  new LevelZeroKernel(module, name);`

File: pmlc/rt/level_zero/level_zero_invocation.h (hunk 13 above)
Comment (YangleiZouIntel) [reply pmlc/rt/level_zero/level_zero_invocation.h]: Ok, these are resources needs to be released. Maybe reword it to something more clear.

File: pmlc/rt/level_zero/level_zero_wrappers.cc [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+  return result;
+}
+
+void level_zero_destroy_execenv(void *invocation) {
+  IVLOG(2, "level_zero_destroy_execenv env=" << invocation);
+  delete static_cast<LevelZeroInvocation *>(invocation);
+  instance = nullptr;
+}
+/*
+void *levelZeroAlloc(void *invocation, size_t bytes) {
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_wrappers.cc]: Please clean up.

File: pmlc/rt/level_zero/level_zero_wrappers.cc [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+  void *result = static_cast<LevelZeroInvocation *>(invocation)
+                     ->enqueueWrite(static_cast<LevelZeroMemory *>(dev), host,
+                                    dependencies);
+  IVLOG(2, "  ->" << result);
+  return result;
+}
+
+void level_zero_wait(uint32_t count, ...) {
+  // No submit created in IR, so call by hand
+  level_zero_submit(instance);
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_wrappers.cc:142]: So we are submitting at every comp.Wait Op (is it 1:1 mapping to level_zero_wait?)? Not sure of the overhead, this may be something we should keep in mind when tuning the performance.

**Comment by YangleiZouIntel** [comment]
Mostly nitpicking. This is great work!

File: plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt (hunk 1 above)
Comment (XinWangIntel) [reply plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:1]: Some of them can pass, i have remove them from the file. Others may fail for stdx lowering, opencl spirv namespace(we just use GLSL.450), IR lowering error...

File: plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt (hunk 2 above)
Comment (XinWangIntel) [reply plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt]: Removed

File: pmlc/BUILD (hunk 3 above)
Comment (XinWangIntel) [reply pmlc/BUILD]: Done

File: pmlc/rt/level_zero/.gdb_history (hunk 4 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/.gdb_history]: Removed

File: pmlc/conversion/comp_to_llvm/passes.td (hunk 5 above)
Comment (XinWangIntel) [reply pmlc/conversion/comp_to_llvm/passes.td]: Removed as we use comp_to_llvm now, but i wonder if we shall keep a separate pass for level zero as we may need some info such as where to add submit and when we can release an event.

File: pmlc/conversion/comp_to_llvm/passes.h (hunk 6 above)
Comment (XinWangIntel) [reply pmlc/conversion/comp_to_llvm/passes.h]: Done

File: pmlc/rt/level_zero/level_zero_device.cc (hunk 7 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_device.cc]: Done

File: pmlc/rt/level_zero/level_zero_invocation.cc (hunk 8 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Done

File: pmlc/rt/level_zero/level_zero_device.cc (hunk 9 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_device.cc]: Removed

File: pmlc/rt/level_zero/level_zero_invocation.cc (hunk 10 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: comp dialect now create xxx_destroy_kernel to release kernel, so remove these.  But not sure how it will change

File: pmlc/rt/level_zero/level_zero_invocation.h (hunk 11 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.h]: Removed

File: pmlc/rt/level_zero/level_zero_invocation.h (hunk 12 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.h:145]: Want to control the time we release resource, as the lifespan of an object is hard to decided until we have a solid pipeline upon.

File: pmlc/rt/level_zero/level_zero_invocation.h (hunk 13 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.h]: Done

File: pmlc/rt/level_zero/level_zero_invocation.cc (hunk 14 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Done

File: pmlc/rt/level_zero/level_zero_invocation.cc (hunk 15 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Removed kernels

File: pmlc/rt/level_zero/level_zero_wrappers.cc (hunk 16 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_wrappers.cc]: Done

File: pmlc/rt/level_zero/level_zero_wrappers.cc (hunk 17 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_wrappers.cc:142]: In L0, each command shall be executed by hand and the command list needs to be reset to be use again,  so do submit before wait to avoid too much submit.  May tune it based on comp dialect.

File: plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp [hunk 18]
```diff
@@ -47,7 +47,7 @@ const std::map<ActivationTypes, std::vector<std::vector<float>>> activationTypes
     {Swish, {{1.0f}}},
     {Mish, {{}}},
     {HSwish, {{}}},
-    {SoftPlus, {{}}}
+    {SoftPlus, {{}}},
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp]: Are these openvino changes related to L0? I think we should make them a separate PR.

File: pmlc/rt/level_zero/level_zero_device.cc [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+                           mlir::ArrayRef<void *>{this});
+}
+
+LevelZeroQueueUser
+LevelZeroDevice::getQueue(ze_command_queue_group_properties_t properties) {
+  // Lock modification of queues vector.
+  std::lock_guard<std::mutex> lock(queuesMutex);
+  for (std::unique_ptr<LevelZeroQueueGuard> &guard : queues) {
+    // Use same kind of queue now, open once open new device
+    if (/*!(guard->getLevelZeroProperties() & properties) ||*/ guard
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_device.cc:88]: Please add a comment or TODO  here, will need to get queue by properties and here we will compare the property struct (including the properties.flag)

File: pmlc/rt/level_zero/level_zero_device.cc (hunk 9 above)
Comment (YangleiZouIntel) [reply pmlc/rt/level_zero/level_zero_device.cc]: Thanks, do you mean removed the comments here or do you mean something else? please check.

File: plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt (hunk 1 above)
Comment (YangleiZouIntel) [reply plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:1]: Thanks, is this level_zero skip list the same as opencl skip list? I assume there is a little difference. Yes most the failed tests are due to spirv ocl ops not added in upstream yet, and the pass loweing to spirv is default to GLSL450 ops.

Comment (XinWangIntel) [reply plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:1]: No, they shall be same,  but  some cases in opencl skip list can pass now (such like Max), so i remove them.  Not appropriate to change opencl skip list in this PR.

File: plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp (hunk 18 above)
Comment (XinWangIntel) [reply plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp]: This from latest plaidml-v1 branch

Comment (YangleiZouIntel) [reply plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp]: I see.

File: pmlc/rt/level_zero/level_zero_device.cc (hunk 19 above)
Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_device.cc:88]: Add comments, we now use same kind of queue, so skip this check, may open when we find better ways to create a queue

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+  std::vector<ze_event_handle_t> zeEvents;
+  std::transform(events.begin(), events.end(), std::back_inserter(zeEvents),
+                 [](const LevelZeroEvent *event) { return event->getEvent(); });
+  for (auto e : zeEvents) {
+    zeEventHostSynchronize(e, UINT64_MAX);
+  }
+}
+
+LevelZeroInvocation::LevelZeroInvocation(LevelZeroDevice *device)
+    : device{device->shared_from_this()}, queueUser(device->getQueue(p)) {
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc:72]: Is the `p` in ` queueUser(device->getQueue(p)` a global vaiable? Can we change it to some default settings as in line 76-78?

Comment (YangleiZouIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc:72]: ok, it is a private member now. It is a little wired, but I think it is ok for now and there is enough comments we can improve in the future.

Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc:72]: Yes, the queue property is not really used now, as we use command queue and command list in L0, we may do complex check when more features are opened.

File: pmlc/rt/level_zero/level_zero_invocation.cc [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+
+void LevelZeroKernel::setArg(unsigned idx, LevelZeroMemory *memory) {
+  void *buf = reinterpret_cast<int64_t *>(memory->getBuffer());
+  lzu::set_argument_value(kernel, idx, sizeof(buf), &buf);
+}
+
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // The usage of gws and lws will depend on their meaning.
```
Comment (YangleiZouIntel) [inline pmlc/rt/level_zero/level_zero_invocation.cc]: <del>I think this comments is good for us. But for PR to plaidml, we can remove this comment.

Comment (YangleiZouIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Maybe it is better to replace this comments by the meaning of gws in lzu::append_launch_function

Comment (YangleiZouIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Please add a new line for unused variable like `(void)lws;` or define a macro like `#define UNUSED(x) [&x]{}()`   then `UNUSED(lws);` to avoid unused warning.

Comment (XinWangIntel) [reply pmlc/rt/level_zero/level_zero_invocation.cc]: Add UNUSED_VARIABLE with (void)(x),  may be save space.     Reword and move comment to enqueueKennel function.

File: pmlc/conversion/comp_to_llvm/passes.td (hunk 5 above)
Comment (YangleiZouIntel) [reply pmlc/conversion/comp_to_llvm/passes.td]: This is a good point. I think we should consider adding a new pass to add submit Ops in the future (and also for vulkan).

Comment (YangleiZouIntel) [reply pmlc/conversion/comp_to_llvm/passes.td]: We definitly will have L0 specific passes including the one discussed here to improve performance.

Comment (XinWangIntel) [reply pmlc/conversion/comp_to_llvm/passes.td]: If they use comp_to_llvm with prefix for all targets, we may need to add a pass before or after that to add more target specific changes

**Comment by XinWangIntel** [comment]
Submit to plaidml and will work on that PR.
